hq/app/schedule/IamAudit.scala

```diff
@@ -43,6 +41,65 @@ object IamAudit extends Logging {
     }.collect { case (awsAccount, Right(report)) => (awsAccount, report) }
   }
 
+  private def sortUsersIntoWarningOrFinalAlerts(users: Seq[VulnerableUser]): (Seq[VulnerableUser], Seq[VulnerableUser]) = {
+    val warningAlerts = users.filter(user => isWarningAlert(createDeadlineIfMissing(user.disableDeadline)))
+    val finalAlerts = users.filter(user => isFinalAlert(createDeadlineIfMissing(user.disableDeadline)))
+    (warningAlerts, finalAlerts)
+  }
+
+  private def createNotifications(warning: Boolean, users: Seq[VulnerableUser], awsAccount: AwsAccount, targets: List[Target]): Seq[IamNotification] = {
+    users.map { user =>
+      val message = if (warning) createWarningMessage(awsAccount, users) else createFinalMessage(awsAccount, users)
+      createNotification(awsAccount, targets :+ Account(awsAccount.accountNumber), message, warningSubject(awsAccount), user.username, createDeadlineIfMissing(user.disableDeadline))
+    }
+  }
+
+  private def createDeadlineIfMissing(date: Option[DateTime]): DateTime = date.getOrElse(DateTime.now.plusDays(iamAlertCadence))
+
+  def isWarningAlert(deadline: DateTime, today: DateTime = DateTime.now): Boolean = {
+    deadline.withTimeAtStartOfDay == today.withTimeAtStartOfDay.plusWeeks(1) ||
+      deadline.withTimeAtStartOfDay == today.withTimeAtStartOfDay.plusWeeks(3)
+  }
+  def isFinalAlert(deadline: DateTime, today: DateTime = DateTime.now): Boolean = deadline.withTimeAtStartOfDay == today.withTimeAtStartOfDay.plusDays(1)
+
+  private def getTargetGroups(report: CredentialReportDisplay, awsAccount: AwsAccount, dynamo: Dynamo): Seq[IAMAlertTargetGroup] = {
+    val vulnerableUsers = findVulnerableUsers(report)
+    val vulnerableUsersToAlert = filterUsersToAlert(vulnerableUsers, awsAccount, dynamo)
+    getNotificationTargetGroups(vulnerableUsersToAlert)
+  }
+
+  private def findVulnerableUsers(report: CredentialReportDisplay): Seq[VulnerableUser] = {
+    outdatedKeysInfo(findOldAccessKeys(report)) ++ missingMfaInfo(findMissingMfa(report))
+  }
+
+  // if the user is not present in dynamo, that means they've never been alerted before, so mark them as ready to be alerted
+  private def filterUsersToAlert(users: Seq[VulnerableUser], awsAccount: AwsAccount, dynamo: Dynamo): Seq[VulnerableUser] = {
+    getExistingDeadlines(users, awsAccount, dynamo).filter { user =>
+      user.disableDeadline match {
+        case Some(u) => isWarningAlert(u) || isFinalAlert(u)
+        case None => true
```

Comment:
Scala experts in my first team at the guardian were of the opinion that pattern matching on options is rarely a good idea so I've sort of never done it. This is pretty readable though!

`user.disableDeadline.exists(u => isWarningAlert(u) || isFinalAlert(u)) || user.disableDeadline.isEmpty` would be another way of writing this